nemo/src/model/rule_model/rule.rs

```diff
@@ -51,27 +51,34 @@ impl Rule {
             .iter()
             .cloned()
             .partition(|literal| literal.is_positive());
-        let negative_variables = negative
-            .iter()
-            .flat_map(|literal| literal.universal_variables())
-            .collect::<HashSet<_>>();
         let positive_varibales = positive
             .iter()
             .flat_map(|literal| literal.universal_variables())
-            .collect();
-        let negative_variables = negative_variables
-            .difference(&positive_varibales)
-            .collect::<Vec<_>>();
+            .collect::<HashSet<&Variable>>();
 
-        if !negative_variables.is_empty() {
-            return Err(ParseError::UnsafeNegatedVariable(
-                negative_variables
-                    .first()
-                    .expect("is not empty here")
-                    .name(),
-            ));
+        let mut unsafe_negative_variables = HashSet::<Variable>::new();
+        for negative_literal in negative {
+            let mut current_unsafe = HashSet::<Variable>::new();
+
+            for variable in negative_literal.variables() {
+                if positive_varibales.contains(variable) {
+                    continue;
+                }
+
+                if unsafe_negative_variables.contains(variable) {
+                    return Err(ParseError::UnsafeVariableInMulltipleNegativeLiterals(
+                        variable.name(),
+                    ));
+                }
+
+                current_unsafe.insert(variable.clone());
+            }
+
+            unsafe_negative_variables.extend(current_unsafe)
         }
 
+        // TODO: Implement something that checks whether unsafe negated variables occurs in multiple places
+
```

Comment:
This is what the code above does, isn't it?

Reply:
Yes, indeed